src/systemctl/systemctl-list-units.c

```diff
@@ -287,20 +287,81 @@ static int get_triggered_units(
         return 0;
 }
 
-static int get_listening(
+struct socket_info {
+        const char *machine;
+        const char* id;
+
+        char* type;
+        char* path;
+
+        /* Note: triggered is a list here, although it almost certainly will always be one
+         * unit. Nevertheless, dbus API allows for multiple values, so let's follow that. */
+        char** triggered;
+};
+
+struct socket_infos {
+        struct socket_info *sockets;
+        size_t n_sockets;
+};
+
+static void socket_infos_done(struct socket_infos *ss) {
+        assert(ss);
+        assert(ss->sockets || ss->n_sockets == 0);
+
+        for (size_t i = 0; i < ss->n_sockets; i++) {
+                struct socket_info *s = &ss->sockets[i];
+
+                free(s->type);
+                free(s->path);
+                strv_free(s->triggered);
+        }
+
+        free(ss->sockets);
+}
+
+static int socket_info_compare(const struct socket_info *a, const struct socket_info *b) {
+        int r;
+
+        assert(a);
+        assert(b);
+
+        r = strcasecmp_ptr(a->machine, b->machine);
+        if (r != 0)
+                return r;
+
+        r = strcmp(a->path, b->path);
```

Comment:
path_compare

Reply:
I thought so, but sometimes it is not a path, e.g. `kobject-uevent 1` or `audit 1`. Do you think still `path_compare()` should be used here?

Reply:
E.g.
```
$ build/systemctl list-sockets --all
LISTEN                                UNIT                            ACTIVATES               
/dev/rfkill                           systemd-rfkill.socket           systemd-rfkill.service
/run/dbus/system_bus_socket           dbus.socket                     dbus-broker.service
/run/dmeventd-client                  dm-event.socket                 dm-event.service
/run/dmeventd-server                  dm-event.socket                 dm-event.service
/run/initctl                          systemd-initctl.socket          systemd-initctl.service
/run/lvm/lvmpolld.socket              lvm2-lvmpolld.socket            lvm2-lvmpolld.service
/run/pcscd/pcscd.comm                 pcscd.socket                    pcscd.service
/run/systemd/coredump                 systemd-coredump.socket         -                       
/run/systemd/journal/dev-log          systemd-journald-dev-log.socket systemd-journald.service
/run/systemd/journal/socket           systemd-journald.socket         systemd-journald.service
/run/systemd/journal/stdout           systemd-journald.socket         systemd-journald.service
/run/systemd/journal/syslog           syslog.socket                   syslog.service
/run/systemd/oom/io.system.ManagedOOM systemd-oomd.socket             systemd-oomd.service
/run/udev/control                     systemd-udevd-control.socket    systemd-udevd.service
audit 1                               systemd-journald-audit.socket   systemd-journald.service
kobject-uevent 1                      systemd-udevd-kernel.socket     systemd-udevd.service

16 sockets listed.
```

Reply:
hmm, good point

Reply:
maybe check if first char is '/', and then use path_compare()? and otherwise strcmp()?